kernel/kernel-api/src/main/java/io/delta/kernel/internal/compaction/LogCompactionWriter.java

```diff
@@ -0,0 +1,162 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.compaction;
+
+import static io.delta.kernel.internal.DeltaErrors.wrapEngineExceptionThrowsIO;
+import static io.delta.kernel.internal.lang.ListUtils.getLast;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.data.FilteredColumnarBatch;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.DeltaLogActionUtils;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.replay.CreateCheckpointIterator;
+import io.delta.kernel.internal.snapshot.LogSegment;
+import io.delta.kernel.internal.util.FileNames.DeltaLogFileType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility for writing out log compactions. */
+public class LogCompactionWriter {
+
+  private static final Logger logger = LoggerFactory.getLogger(LogCompactionWriter.class);
+
+  private final Path logPath;
+  private final long startVersion;
+  private final long endVersion;
+  private final long minFileRetentionTimestampMillis;
+
+  public LogCompactionWriter(
+      Path logPath, long startVersion, long endVersion, long minFileRetentionTimestampMillis) {
+    this.logPath = requireNonNull(logPath);
+    this.startVersion = startVersion;
+    this.endVersion = endVersion;
+    this.minFileRetentionTimestampMillis = minFileRetentionTimestampMillis;
+  }
+
+  public void writeLogCompactionFile(Engine engine) throws IOException {
+    String fileName = String.format("%020d.%020d.compacted.json", startVersion, endVersion);
+    Path compactedPath = new Path(logPath, fileName);
+
+    logger.info(
+        "Writing compacted log file for versions {} to {} to path: {}",
+        startVersion,
+        endVersion,
+        compactedPath);
+
+    final long startTimeMillis = System.currentTimeMillis();
+    final List<FileStatus> deltas =
+        DeltaLogActionUtils.listDeltaLogFilesAsIter(
+                engine,
+                new HashSet<>(Arrays.asList(DeltaLogFileType.COMMIT)),
+                logPath,
+                startVersion,
+                Optional.of(endVersion),
+                true /* mustBeRecreatable */)
```

Comment:
You could _probably_ make this `false`?

When we are loading a `Snapshot` -- we _need_ this to be true.

Here, our goal is to write a log compaction file for, say, versions 100 to 110. If someone just went and did metadata cleanup ... and deleted the checkpoint at version 95 ... that's probably not out fault ...

It's also like ... if we are going to commit version 100 .. and someone just came and deleted 99 ... that's also not our problem right now.

Food for thought -- open to your input.

Reply:
Yeah similar thought, I don't think there's any reason for it to need to be recreatable?

Reply:
yeah agree, I hadn't thought too much about this and was being cautious, but on reflection I think you're right